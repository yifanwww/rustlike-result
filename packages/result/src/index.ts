export * from './factory';
export * from './factoryAsync';
export type * from './Result';
export type * from './ResultAsync';
export * from './resultify';
export type * from './ResultJson';
export * from './RustlikeResult';
export type { ResultType } from './types.internal';
