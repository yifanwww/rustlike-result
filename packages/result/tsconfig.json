{
    "extends": "../../configs/tsconfigs/tsconfig.base.json",
    "compilerOptions": {
        "incremental": false,
        "noEmit": true,
        "rootDir": "."
    },
    "include": ["src/**/*"],
    "exclude": []
}
