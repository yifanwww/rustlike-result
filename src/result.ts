import type { IResult, Optional } from './types';

type ResultType = 'ok' | 'err';

/**
 * `Result` is a type that represents either success (Ok) or failure (Err).
 *
 * Handle Errors with the `Result` type.
 *
 * `Result<T, E>` is used for returning and propagating errors.
 * It is a type with two kinds:
 * - `Result.Ok(T)`, representing success and containing a value
 * - `Result.Err(E)`, representing error and containing an error value.
 *
 * ref:
 * - https://doc.rust-lang.org/std/result/index.html
 * - https://doc.rust-lang.org/std/result/enum.Result.html
 */
export class Result<T, E> implements IResult<T, E> {
    private readonly _type: ResultType;
    private _value?: T;
    private _error?: E;

    private constructor(type: 'ok', value: T);
    private constructor(type: 'err', error: E);
    private constructor(type: ResultType, value: T | E) {
        if (type === 'ok') {
            this._type = 'ok';
            this._value = value as T;
        } else {
            this._type = 'err';
            this._error = value as E;
        }
    }

    /**
     * Contains the success value.
     */
    static Ok<T, E = never>(value: T): IResult<T, E> {
        return new Result<T, E>('ok', value);
    }

    /**
     * Contains the error value.
     */
    static Err<E, T = never>(error: E): IResult<T, E> {
        return new Result<T, E>('err', error);
    }

    /**
     * Returns `true` if the result is `Ok`.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.is_ok
     */
    isOk(): boolean {
        return this._type === 'ok';
    }

    /**
     * Returns `true` if the result is `Ok` and the value inside of it matches a predicate.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.is_ok_and
     */
    isOkAnd(fn: (value: T) => boolean): boolean {
        return this.isOk() && fn(this._value!);
    }

    /**
     * Returns `true` if the result is `Err`.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.is_err
     */
    isErr(): boolean {
        return this._type === 'err';
    }

    /**
     * Returns `true` if the result is `Err` and the value inside of it matches a predicate.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.is_err_and
     */
    isErrAnd(fn: (err: E) => boolean): boolean {
        return this.isErr() && fn(this._error!);
    }

    /**
     * Converts from `Result<T, E>` to `Optional<T>` and discarding the error, if any.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.ok
     */
    ok(): Optional<T> {
        return this.isOk() ? this._value : undefined;
    }

    /**
     * Converts from `Result<T, E>` to `Optional<E>` and discarding the success value, if any.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.err
     */
    err(): Optional<E> {
        return this.isOk() ? undefined : this._error;
    }

    /**
     * Maps a `Result<T, E>` to `Result<U, E>` by applying a function to a contained `Ok` value,
     * leaving an `Err` value untouched.
     *
     * This function can be used to compose the results of two functions.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.map
     */
    map<U>(op: (value: T) => U): IResult<U, E> {
        return this.isOk() ? Result.Ok(op(this._value!)) : Result.Err(this._error!);
    }

    /**
     * Returns the provided `fallback` (if `Err`), or applies a function to the contained value (if `Ok`).
     *
     * Arguments passed to `mapOr` are eagerly evaluated;
     * if you are passing the result of a function call,
     * it is recommended to use `mapOrElse`, which is lazily evaluated.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or
     */
    mapOr<U>(fallback: U, map: (value: T) => U): U {
        return this.isOk() ? map(this._value!) : fallback;
    }

    /**
     * Maps a `Result<T, E>` to `U` by applying fallback function `fallback` to a contained `Err` value,
     * or function `map` to a contained `Ok` value.
     *
     * This function can be used to unpack a successful result while handling an error.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or_else
     */
    mapOrElse<U>(fallback: (err: E) => U, map: (value: T) => U): U {
        return this.isOk() ? map(this._value!) : fallback(this._error!);
    }

    /**
     * Maps a `Result<T, E>` to `Result<T, F>` by applying a function to a contained `Err` value,
     * leaving an `Ok` value untouched.
     *
     * This function can be used to pass through a successful result while handling an error.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err
     */
    mapErr<F>(op: (err: E) => F): IResult<T, F> {
        return this.isOk() ? Result.Ok(this._value!) : Result.Err(op(this._error!));
    }

    private _unwrapFailed(msg: string, error: unknown): never {
        throw new Error(`${msg}: ${String(error)}`);
    }

    /**
     * Returns the contained `Ok` value.
     *
     * Throws Error if itself is `Err`, with a error message provided by the `Err`'s value.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.expect
     */
    expect(msg: string): T {
        return this.isOk() ? this._value! : this._unwrapFailed(msg, this._error!);
    }

    /**
     * Returns the contained `Ok` value.
     *
     * Throws Error if itself is `Err`, with a error message provided by the `Err`'s value.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap
     */
    unwrap(): T {
        if (this.isOk()) return this._value!;
        throw new Error(String(this._error!));
    }

    /**
     * Returns the contained `Err` value.
     *
     * Throws Error if itself is `Err`, with a error message provided by the `Err`'s value.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.expect_err
     */
    expectErr(msg: string): E {
        return this.isErr() ? this._error! : this._unwrapFailed(msg, this._value!);
    }

    /**
     * Returns the contained `Err` value.
     *
     * Throws Error if itself is `Ok`, with a error message provided by the `Ok`'s value.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_err
     */
    unwrapErr(): E {
        if (this.isErr()) return this._error!;
        throw new Error(String(this._value!));
    }

    /**
     * Returns the contained `Ok` value or a provided default.
     *
     * Arguments passed to `unwrapOr` are eagerly evaluated;
     * if you are passing the result of a function call,
     * it is recommended to use `unwrapOrElse`, which is lazily evaluated.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or
     */
    unwrapOr(fallback: T): T {
        return this.isOk() ? this._value! : fallback;
    }

    /**
     * Returns the contained `Ok` value or computes it from a closure.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or_else
     */
    unwrapOrElse(op: (err: E) => T): T {
        return this.isOk() ? this._value! : op(this._error!);
    }

    /**
     * Returns `res` if itself is `Ok`, otherwise returns the `Err` value of itself.
     *
     * Arguments passed to `and` are eagerly evaluated;
     * if you are passing the result of a function call, it is recommended to use `andThen`, which is lazily evaluated.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.and
     */
    and<U>(res: IResult<U, E>): IResult<U, E> {
        return this.isOk() ? res : Result.Err(this._error!);
    }

    /**
     * Calls `op` if itself is `Ok`, otherwise returns the `Err` value of itself.
     *
     * This function can be used for control flow based on Result values.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.and_then
     */
    andThen<U>(op: (value: T) => IResult<U, E>): IResult<U, E> {
        return this.isOk() ? op(this._value!) : Result.Err(this._error!);
    }

    /**
     * Returns `res` if itself is `Err`, otherwise returns the `Ok` value of itself.
     *
     * Arguments passed to `or` are eagerly evaluated;
     * if you are passing the result of a function call, it is recommended to use `orElse`, which is lazily evaluated.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.or
     */
    or<F>(res: IResult<T, F>): IResult<T, F> {
        return this.isOk() ? Result.Ok(this._value!) : res;
    }

    /**
     * Calls `op` if the result is `Err`, otherwise returns the `Ok` value of self.
     *
     * This function can be used for control flow based on result values.
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else
     */
    orElse<F>(op: (error: E) => IResult<T, F>): IResult<T, F> {
        return this.isOk() ? Result.Ok(this._value!) : op(this._error!);
    }

    /**
     * Transposes a `Result` of an optional value into an optional of a `Result`.
     * - `Ok(undefined | null)` will be mapped to `undefined`.
     * - `Ok(_)` (`Ok(Some(_))` in Rust) will be mapped to `Ok(_)` (`Some(Ok(_))` in Rust).
     * - `Err(_)` will be mapped to `Err(_)` (`Some(Err(_))` in Rust).
     *
     * ref: https://doc.rust-lang.org/std/result/enum.Result.html#method.transpose
     */
    transpose(): Optional<IResult<T & NonNullable<unknown>, E>> {
        if (this.isOk()) {
            return this._value === undefined || this._value === null ? undefined : Result.Ok(this._value);
        }
        return Result.Err(this._error!);
    }

    private static _equal(self: unknown, other: unknown): boolean {
        const isSelfResult = self instanceof Result;
        const isOtherResult = other instanceof Result;

        if (isSelfResult && isOtherResult) {
            const isOk = self.isOk();
            if (isOk !== other.isOk()) return false;
            return isOk ? Result._equal(self._value, other._value) : Result._equal(self._error, other._error);
        }

        return self === other || (Number.isNaN(self) && Number.isNaN(other));
    }

    /**
     * Returns `true` if `self` equals to `other`.
     */
    equal(other: Result<T, E>): boolean {
        const isOk = this.isOk();
        if (isOk !== other.isOk()) return false;
        return isOk ? Result._equal(this._value, other._value!) : Result._equal(this._error, other._error!);
    }
}

export function Ok<T>(value: T): IResult<T, never>;
export function Ok<T, E>(value: T): IResult<T, E>;
export function Ok<T, E>(value: T): IResult<T, E> {
    return Result.Ok(value);
}

export function Err<E>(error: E): IResult<never, E>;
export function Err<T, E>(error: E): IResult<T, E>;
export function Err<T, E>(error: E): IResult<T, E> {
    return Result.Err(error);
}
