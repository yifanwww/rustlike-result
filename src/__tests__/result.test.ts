import { Err, Ok, Result } from '../result';
import type { IResult } from '../types';

function op1(): IResult<number, string> {
    return Ok(666);
}

function op2(): IResult<number, string> {
    return Err('sadface');
}

describe(`Test static method \`${Result.name}.${Result.Ok.name}\``, () => {
    it('should create Ok result', () => {
        expect(Result.Ok(1)).toMatchSnapshot();
    });
});

describe(`Test static method \`${Result.name}.${Result.Err.name}\``, () => {
    it('should create Err result', () => {
        expect(Result.Err('Some error message')).toMatchSnapshot();
    });
});

describe(`Test fn \`${Ok.name}\``, () => {
    it('should create Ok result', () => {
        expect(Ok(1)).toMatchSnapshot();
    });
});

describe(`Test fn \`${Err.name}\``, () => {
    it('should create Err result', () => {
        expect(Err('Some error message')).toMatchSnapshot();
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.isOk.name}\``, () => {
    it('should return if itself is `Ok`', () => {
        expect(Ok(1).isOk()).toBe(true);
        expect(Err('Some error message').isOk()).toBe(false);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.isOkAnd.name}\``, () => {
    const fnFactory = () => jest.fn((num: number) => num > 1);

    it('should return if itself is `Ok` and the value inside of it matches a predicate', () => {
        const fn = fnFactory();
        expect(Ok(2).isOkAnd(fn)).toBe(true);
        expect(Ok(0).isOkAnd(fn)).toBe(false);
        expect(Err('Some error message').isOkAnd(fn)).toBe(false);
    });

    it('should call fn only once if itself is `Ok`', () => {
        const fn = fnFactory();
        expect(fn).toHaveBeenCalledTimes(0);
        Ok(2).isOkAnd(fn);
        expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should not call fn if itself is `Err`', () => {
        const fn = fnFactory();
        expect(fn).toHaveBeenCalledTimes(0);
        Err('Some error message').isOkAnd(fn);
        expect(fn).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.isErr.name}\``, () => {
    it('should return if itself is `Err`', () => {
        expect(Ok(1).isErr()).toBe(false);
        expect(Err('Some error message').isErr()).toBe(true);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.isErrAnd.name}\``, () => {
    enum ErrorKind {
        NOT_FOUND,
        PERMISSION_DENIED,
    }

    const fnFactory = () => jest.fn((err: ErrorKind) => err === ErrorKind.NOT_FOUND);

    it('should return if itself is `Err` and the value inside of it matches a predicate', () => {
        const fn = fnFactory();
        expect(Err(ErrorKind.NOT_FOUND).isErrAnd(fn)).toBe(true);
        expect(Err(ErrorKind.PERMISSION_DENIED).isErrAnd(fn)).toBe(false);
        expect(Ok(123).isErrAnd(fn)).toBe(false);
    });

    it('should call fn only once if itself is `Err`', () => {
        const fn = fnFactory();
        expect(fn).toHaveBeenCalledTimes(0);
        Err(ErrorKind.NOT_FOUND).isErrAnd(fn);
        expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should not call fn if itself is `Ok`', () => {
        const fn = fnFactory();
        expect(fn).toHaveBeenCalledTimes(0);
        Ok(123).isErrAnd(fn);
        expect(fn).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.ok.name}\``, () => {
    it('should convert itself to an optional value', () => {
        expect(Ok(1).ok()).toBe(1);
        expect(Err('Some error message').ok()).toBeUndefined();
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.err.name}\``, () => {
    it('should convert itself to an optional error', () => {
        expect(Ok(1).err()).toBeUndefined();
        expect(Err('Some error message').err()).toBe('Some error message');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.map.name}\``, () => {
    const mapFactory = () => jest.fn((num: number) => String(num));

    it('should map itself to another result', () => {
        const map = mapFactory();
        expect(Ok(1).map(map).ok()).toBe('1');
        expect(Err('Some error message').map(map).ok()).toBeUndefined();
    });

    it('should call map fn only once if itself is `Ok`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Ok(1).map(map);
        expect(map).toHaveBeenCalledTimes(1);
    });

    it('should not call map fn if itself is `Err`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Err('Some error message').map(map);
        expect(map).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.mapOr.name}\``, () => {
    const mapFactory = () => jest.fn((num: number) => num * 2);

    it('should map itself to another result', () => {
        const map = mapFactory();
        expect(Ok(1).mapOr(-1, map)).toBe(2);
        expect(Err('Some error message').mapOr(-1, map)).toBe(-1);
    });

    it('should call map fn only once if itself is `Ok`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Ok(1).mapOr(-1, map);
        expect(map).toHaveBeenCalledTimes(1);
    });

    it('should not call map fn if itself is `Err`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Err('Some error message').mapOr(-1, map);
        expect(map).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.mapOrElse.name}\``, () => {
    const mapFactory = () => jest.fn(jest.fn((num: number) => String(num)));
    const fallbackFactory = () => jest.fn(jest.fn((str: string) => str));

    it('should map itself to another value', () => {
        const map = mapFactory();
        const fallback = fallbackFactory();
        expect(Ok(1).mapOrElse(fallback, map)).toBe('1');
        expect(Err('Some error message').mapOrElse(fallback, map)).toBe('Some error message');
    });

    it('should call map fn only once if itself is `Ok`', () => {
        const map = mapFactory();
        const fallback = fallbackFactory();

        expect(map).toHaveBeenCalledTimes(0);
        expect(fallback).toHaveBeenCalledTimes(0);
        Ok(1).mapOrElse(fallback, map);
        expect(map).toHaveBeenCalledTimes(1);
        expect(fallback).toHaveBeenCalledTimes(0);
    });

    it('should call fallback fn only once if itself is `Err`', () => {
        const map = mapFactory();
        const fallback = fallbackFactory();

        expect(map).toHaveBeenCalledTimes(0);
        expect(fallback).toHaveBeenCalledTimes(0);
        Err('Some error message').mapOrElse(fallback, map);
        expect(map).toHaveBeenCalledTimes(0);
        expect(fallback).toHaveBeenCalledTimes(1);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.mapErr.name}\``, () => {
    const mapFactory = () => jest.fn((num: number) => `error code: ${num}`);

    it('should map itself to another result', () => {
        const map = mapFactory();
        expect(Ok(1).mapErr(map).ok()).toBe(1);
        expect(Err(2).mapErr(map).err()).toBe('error code: 2');
    });

    it('should call map fn only once if itself is `Err`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Err(2).mapErr(map);
        expect(map).toHaveBeenCalledTimes(1);
    });

    it('should not call map fn if itself is `Ok`', () => {
        const map = mapFactory();
        expect(map).toHaveBeenCalledTimes(0);
        Ok(1).mapErr(map);
        expect(map).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.expect.name}\``, () => {
    it('should unwrap itself to get the contained `Ok` value', () => {
        expect(Ok(1).expect('Operation type should be correct')).toBe(1);
        expect(() => Err(2).expect('Operation type should be correct')).toThrow('Operation type should be correct: 2');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.unwrap.name}\``, () => {
    it('should unwrap itself to get the contained `Ok` value', () => {
        expect(Ok(1).unwrap()).toBe(1);
        expect(() => Err('Some error message').unwrap()).toThrow('Some error message');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.expectErr.name}\``, () => {
    it('should unwrap itself to get the contained `Err` value', () => {
        expect(Err('Some error message').expectErr('Testing expectErr')).toBe('Some error message');
        expect(() => Ok(1).expectErr('Testing expectErr')).toThrow('Testing expectErr: 1');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.unwrapErr.name}\``, () => {
    it('should unwrap itself to get the contained `Err` value', () => {
        expect(Err('Some error message').unwrapErr()).toBe('Some error message');
        expect(() => Ok(1).unwrapErr()).toThrow('1');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.unwrapOr.name}\``, () => {
    it('should unwrap itself to get the contained `Ok` value or the provided value', () => {
        const ok = Ok<number, string>(100);
        const okErr = Err<number, string>('Err');

        expect(ok.unwrapOr(50)).toBe(100);
        expect(okErr.unwrapOr(50)).toBe(50);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.unwrapOrElse.name}\``, () => {
    it('should unwrap itself to get the contained `Ok` value or computes it from a closure', () => {
        const handler = (msg: string) => {
            if (msg === 'I got this.') return 50;
            throw new Error('BadBad');
        };

        const ok = Ok<number, string>(100);
        const okErr = Err<number, string>('I got this.');

        expect(ok.unwrapOrElse(handler)).toBe(100);
        expect(okErr.unwrapOrElse(handler)).toBe(50);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.and.name}\``, () => {
    it('should return `res`', () => {
        expect(op1().and(Ok(667)).ok()).toBe(667);
        expect(op1().and(Err('bad')).err()).toBe('bad');
    });

    it('should return the `Err` result', () => {
        expect(op2().and(Ok(667)).err()).toBe('sadface');
        expect(op2().and(Err('bad')).err()).toBe('sadface');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.andThen.name}\``, () => {
    it('should return `res`', () => {
        expect(
            op1()
                .andThen((num) => Ok(num + 1))
                .ok(),
        ).toBe(667);

        expect(
            op1()
                .andThen(() => Err('bad'))
                .err(),
        ).toBe('bad');
    });

    it('should return the `Err` result', () => {
        expect(
            op2()
                .andThen((num) => Ok(num + 1))
                .err(),
        ).toBe('sadface');

        expect(
            op2()
                .andThen(() => Err('bad'))
                .err(),
        ).toBe('sadface');
    });

    it('should call op fn only once if itself is `Ok`', () => {
        const fn = jest.fn((num: number) => Ok(num + 1));
        expect(fn).toHaveBeenCalledTimes(0);
        op1().andThen(fn);
        expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should not call op fn if itself is `Err`', () => {
        const fn = jest.fn((num: number) => Ok(num + 1));
        expect(fn).toHaveBeenCalledTimes(0);
        op2().andThen(fn);
        expect(fn).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.or.name}\``, () => {
    it('should return the `Ok` result', () => {
        expect(op1().or(Ok(667)).ok()).toBe(666);
        expect(op1().or(Err('bad')).ok()).toBe(666);
    });

    it('should return `res`', () => {
        expect(op2().or(Ok(667)).ok()).toBe(667);
        expect(op2().or(Err('bad')).err()).toBe('bad');
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.orElse.name}\``, () => {
    it('should return the `Ok` result', () => {
        expect(
            op1()
                .orElse(() => Ok(667))
                .ok(),
        ).toBe(666);

        expect(
            op1()
                .orElse((err) => Err(err))
                .ok(),
        ).toBe(666);
    });

    it('should return `res`', () => {
        expect(
            op2()
                .orElse(() => Ok(667))
                .ok(),
        ).toBe(667);

        expect(
            op2()
                .orElse((err) => Err(err))
                .err(),
        ).toBe('sadface');
    });

    it('should call op fn only once if itself is `Err`', () => {
        const fn = jest.fn((err: string) => Err(err));
        expect(fn).toHaveBeenCalledTimes(0);
        op2().orElse(fn);
        expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should not call op fn if itself is `Ok`', () => {
        const fn = jest.fn((err: string) => Err(err));
        expect(fn).toHaveBeenCalledTimes(0);
        op1().orElse(fn);
        expect(fn).toHaveBeenCalledTimes(0);
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.transpose.name}\``, () => {
    it('should transpose itself to an optional of a `Result`', () => {
        expect(Ok<number | undefined | null, string>(1).transpose()!.ok()).toBe(1);
        expect(Ok<number | undefined | null, string>(undefined).transpose()).toBeUndefined();
        expect(Ok<number | undefined | null, string>(null).transpose()).toBeUndefined();
        expect(Err<number | undefined | null, string>('Some error message').transpose()!.err()).toBe(
            'Some error message',
        );
    });
});

describe(`Test method \`${Result.name}.prototype.${Result.prototype.equal.name}\``, () => {
    it('should check if itself equals to another result', () => {
        expect(Ok(1).equal(Ok(1))).toBe(true);
        expect(Ok(NaN).equal(Ok(NaN))).toBe(true);
        expect(Ok(Ok(1)).equal(Ok(Ok(1)))).toBe(true);
        expect(Ok(Err('Some error message')).equal(Ok(Err('Some error message')))).toBe(true);
        expect(Err('Some error message').equal(Err('Some error message'))).toBe(true);
        expect(Err(Err('Some error message')).equal(Err(Err('Some error message')))).toBe(true);
        expect(Err(Ok(1)).equal(Err(Ok(1)))).toBe(true);

        expect(Ok(1).equal(Ok(2))).toBe(false);
        expect(Ok(Ok(1)).equal(Ok(Ok(2)))).toBe(false);
        expect(Ok(Err<number, string>('Some error message')).equal(Ok(Ok(1)))).toBe(false);
        expect(Ok(Ok<number, string>(1)).equal(Ok(Err('Some error message')))).toBe(false);
        expect(Err('Some error message 1').equal(Err('Some error message 2'))).toBe(false);
        expect(Err(Err('Some error message 1')).equal(Err(Err('Some error message 2')))).toBe(false);
        expect(Err(Ok<number, string>(1)).equal(Err(Err('Some error message')))).toBe(false);
        expect(Err(Err<number, string>('Some error message')).equal(Err(Ok(1)))).toBe(false);
        expect(Ok<number, string>(1).equal(Err('Some error message'))).toBe(false);
    });
});
