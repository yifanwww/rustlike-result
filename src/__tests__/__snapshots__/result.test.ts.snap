// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test fn \`Err\` should create Err result 1`] = `
Result {
  "_error": "Some error message",
  "_type": "err",
  "_value": undefined,
}
`;

exports[`Test fn \`Ok\` should create Ok result 1`] = `
Result {
  "_error": undefined,
  "_type": "ok",
  "_value": 1,
}
`;

exports[`Test static method \`Result.Err\` should create Err result 1`] = `
Result {
  "_error": "Some error message",
  "_type": "err",
  "_value": undefined,
}
`;

exports[`Test static method \`Result.Ok\` should create Ok result 1`] = `
Result {
  "_error": undefined,
  "_type": "ok",
  "_value": 1,
}
`;
